doc/indexing.rst

```diff
@@ -648,28 +656,43 @@ dimensions or use the ellipsis in the ``loc`` specifier, e.g. in the example
 above, ``mda.loc[{'one': 'a', 'two': 0}, :]`` or ``mda.loc[('a', 0), ...]``.
 
 
-.. _xarray_indexing_rules:
+.. _indexing.rules:
+
+Indexing rules
+--------------
+
+Here we describe the full rules xarray uses for vectorized indexing. Note that
+this is for the purposes of explanation: for the sake of efficiency and to
+support various backends, the actual implementation is different.
 
-xarray indexing rules
----------------------
+0. (Only for label based indexing.) Look up positional indexes along each
+   dimension from the corresponding :py:class:`pandas.Index`.
 
-The detailed indexing scheme in xarray is as follows.
-(Note that it is for the explanation purpose and the actual implementation is differ.)
+1. A full slice object ``:`` is inserted for each dimension without an indexer.
 
-0. (Only for label based indexing.) Look up positional indexes along each dimension based on :py:class:`pandas.Index`.
+2. ``slice`` objects are converted into arrays, given by
+   ``np.arange(*slice.indices(...))``.
 
-1. ``slice`` is converted to an array, such that ``np.arange(*slice.indices(...))``.
+3. Assume dimension names for array indexers without dimensions, such as
+   ``np.ndarray`` and ``list``, from the dimensions to be indexed along.
+   For example, ``v.isel(x=[0, 1])`` is understood as
+   ``v.isel(x=xr.DataArray([0, 1], dims=['x']))``.
 
-2. Assume dimension names of array indexers without dimension, such as ``np.ndarray`` and ``list``, from the dimensions to be indexed along. For example, ``v.isel(x=[0, 1])`` is understood as ``v.isel(x=xr.DataArray([0, 1], dims=['x']))``.
+4. For each variable in a ``Dataset`` or  ``DataArray`` (the array and its
+   coordinates):
 
-3. Broadcast all the indexers based on their dimension names (see :ref:`compute.broadcasting` for our name-based broadcasting).
+   a. Broadcast all relevant indexers based on their dimension names
+      (see :ref:`compute.broadcasting` for full details).
 
-4. Index the object by the broadcasted indexers.
+   b. Index the underling array by the broadcast indexers, using NumPy's
+      advanced indexing rules.
 
-5. If an indexer-DataArray has coordinates, attached them to the indexed object.
+5. If any indexer DataArray has coordinates and no coordinate with the
+   same name exists, attach them to the indexed object.
 
 .. note::
 
-  + There should not be a conflict between the coordinates of indexer- and indexed- DataArrays. In v.0.10.0, xarray raises ``FutureWarning`` if there is such a conflict, but in the next major release, it will raise an Error.
+  - ``IndexError`` is raised if there is a conflict between dimension
+     coordinates of indexer- and indexed- DataArrays.
```

Comment:
Should we raise `IndexError` for coordinate conflict?
The current code just neglects the conflicted coordinates.

Reply:
I think we do want to raise an error here, but as you note this is not the case currently. Let's discuss this over in the main issue (where I also just raised this concern)

Reply:
I removed this line for now. Let's add it back in later when we settle on how it should work.